File: agent/internal/noise/interop_test.go
// agent/internal/noise/interop_test.go
package noise

import (
	"crypto/ecdh"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// InteropVector captures the full handshake transcript for JS verification.
type InteropVector struct {
	InitiatorEphemeralPriv string `json:"initiator_ephemeral_priv"` // hex
	ResponderEphemeralPriv string `json:"responder_ephemeral_priv"`
	InitiatorStaticPriv    string `json:"initiator_static_priv"`
	ResponderStaticPriv    string `json:"responder_static_priv"`
	Msg1                   string `json:"msg1"`
	Msg2                   string `json:"msg2"`
	Msg3                   string `json:"msg3"`
	K1                     string `json:"k1"` // initiator→responder post-handshake key
	K2                     string `json:"k2"` // responder→initiator post-handshake key
	// A sample encrypted message from each direction
	SamplePlaintext   string `json:"sample_plaintext"`
	SampleCtInitiator string `json:"sample_ct_initiator"` // iSend.Encrypt(sample)
	SampleCtResponder string `json:"sample_ct_responder"` // rSend.Encrypt(sample)
}

// privateKeyFromHex imports a P-256 private key from a 32-byte hex-encoded scalar.
func privateKeyFromHex(t *testing.T, hexStr string) *ecdh.PrivateKey {
	t.Helper()
	raw, err := hex.DecodeString(hexStr)
	if err != nil {
		t.Fatalf("hex decode: %v", err)
	}
	priv, err := ecdh.P256().NewPrivateKey(raw)
	if err != nil {
		t.Fatalf("import private key: %v", err)
	}
	return priv
}

// TestGenerateInteropVectors generates testdata/interop_vector.json.
// Run with: go test -run TestGenerateInteropVectors ./internal/noise/
func TestGenerateInteropVectors(t *testing.T) {
	// Fixed P-256 private key scalars (32 bytes each, hex-encoded).
	// These are arbitrary values in [1, curve_order); verified to be valid P-256 scalars.
	const (
		iEPrivHex = "0102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f20"
		rEPrivHex = "2122232425262728292a2b2c2d2e2f303132333435363738393a3b3c3d3e3f40"
		iSPrivHex = "4142434445464748494a4b4c4d4e4f505152535455565758595a5b5c5d5e5f60"
		rSPrivHex = "6162636465666768696a6b6c6d6e6f707172737475767778797a7b7c7d7e7f00"
	)

	iEPriv := privateKeyFromHex(t, iEPrivHex)
	rEPriv := privateKeyFromHex(t, rEPrivHex)
	iSPriv := privateKeyFromHex(t, iSPrivHex)
	rSPriv := privateKeyFromHex(t, rSPrivHex)

	// Build initiator and responder with injected keys
	h, ck := initialize()
	initiator := &NoiseXX{role: roleInitiator, h: h, ck: ck, sPriv: iSPriv, sPub: iSPriv.PublicKey().Bytes()}
	h2, ck2 := initialize()
	responder := &NoiseXX{role: roleResponder, h: h2, ck: ck2, sPriv: rSPriv, sPub: rSPriv.PublicKey().Bytes()}

	// Inject deterministic ephemeral keys (override generateKeypair in WriteMessage1/2)
	initiator.ePriv = iEPriv
	initiator.ePub = iEPriv.PublicKey().Bytes()
	initiator.h = mixHash(initiator.h, initiator.ePub) // simulate WriteMessage1 token processing
	initiator.phase = phaseMsg1Sent

	msg1 := initiator.ePub // 65 bytes

	if err := responder.ReadMessage1(msg1); err != nil {
		t.Fatalf("ReadMessage1: %v", err)
	}

	// Inject responder ephemeral
	responder.ePriv = rEPriv
	responder.ePub = rEPriv.PublicKey().Bytes()

	// Re-run WriteMessage2 logic with injected ephemeral
	// (call the method but it will re-generate ephemeral — we need to intercept)
	// Instead, reconstruct manually using the same steps as WriteMessage2
	responder.h = mixHash(responder.h, responder.ePub)
	msg2buf := make([]byte, 0, 162)
	msg2buf = append(msg2buf, responder.ePub...)

	rEPubKey, _ := importPublicKey(responder.rEPub)
	ee, _ := dhP256(responder.ePriv, rEPubKey)
	responder.mixKey(ee)
	encS, _ := responder.encryptAndHash(responder.sPub)
	msg2buf = append(msg2buf, encS...)
	rEPubKey2, _ := importPublicKey(responder.rEPub)
	es, _ := dhP256(responder.sPriv, rEPubKey2)
	responder.mixKey(es)
	tag, _ := responder.encryptAndHash([]byte{})
	msg2buf = append(msg2buf, tag...)
	msg2 := msg2buf
	responder.phase = phaseMsg2Sent // CRITICAL: must set phase before ReadMessage3 will work

	if err := initiator.ReadMessage2(msg2); err != nil {
		t.Fatalf("ReadMessage2: %v", err)
	}

	// Inject initiator static for WriteMessage3
	initiator.sPriv = iSPriv
	initiator.sPub = iSPriv.PublicKey().Bytes()
	msg3, err := initiator.WriteMessage3()
	if err != nil {
		t.Fatalf("WriteMessage3: %v", err)
	}
	if err := responder.ReadMessage3(msg3); err != nil {
		t.Fatalf("ReadMessage3: %v", err)
	}

	iSend, iRecv := initiator.Split()
	rSend, rRecv := responder.Split()

	const samplePlaintext = "sharebridge-interop-test"
	sampleCTi, _ := iSend.encryptWithAd(nil, []byte(samplePlaintext))
	sampleCTr, _ := rSend.encryptWithAd(nil, []byte(samplePlaintext))

	// Verify cross-decryption before writing the file
	pt1, err := rRecv.decryptWithAd(nil, sampleCTi)
	if err != nil || string(pt1) != samplePlaintext {
		t.Fatalf("cross-decrypt i->r failed: %v", err)
	}
	pt2, err := iRecv.decryptWithAd(nil, sampleCTr)
	if err != nil || string(pt2) != samplePlaintext {
		t.Fatalf("cross-decrypt r->i failed: %v", err)
	}

	vec := InteropVector{
		InitiatorEphemeralPriv: iEPrivHex,
		ResponderEphemeralPriv: rEPrivHex,
		InitiatorStaticPriv:    iSPrivHex,
		ResponderStaticPriv:    rSPrivHex,
		Msg1:                   hex.EncodeToString(msg1),
		Msg2:                   hex.EncodeToString(msg2),
		Msg3:                   hex.EncodeToString(msg3),
		K1:                     hex.EncodeToString(iSend.key[:]),
		K2:                     hex.EncodeToString(rSend.key[:]),
		SamplePlaintext:        samplePlaintext,
		SampleCtInitiator:      hex.EncodeToString(sampleCTi),
		SampleCtResponder:      hex.EncodeToString(sampleCTr),
	}

	if err := os.MkdirAll("testdata", 0755); err != nil {
		t.Fatalf("mkdir testdata: %v", err)
	}
	out, _ := json.MarshalIndent(vec, "", "  ")
	if err := os.WriteFile(filepath.Join("testdata", "interop_vector.json"), out, 0644); err != nil {
		t.Fatalf("write vector: %v", err)
	}
	t.Logf("interop_vector.json written (%d bytes)", len(out))
}

File: signaling-server/web/noise-p256/interop.test.js
// signaling-server/web/noise-p256/interop.test.js
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { readFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import { dirname, join } from 'node:path'

const __dirname = dirname(fileURLToPath(import.meta.url))
const vectorPath = join(__dirname, '../../../agent/internal/noise/testdata/interop_vector.json')
const vec = JSON.parse(readFileSync(vectorPath, 'utf8'))

function fromHex(hex) {
  const bytes = new Uint8Array(hex.length / 2)
  for (let i = 0; i < bytes.length; i++) bytes[i] = parseInt(hex.slice(i*2, i*2+2), 16)
  return bytes
}

import { CipherState } from './cipher.js'

test('interop: JS can decrypt Go-generated ciphertext (initiator->responder)', async () => {
  // k1 is initiator->responder key; responder uses k1 to decrypt
  const k1 = fromHex(vec.k1)
  const rRecv = new CipherState(k1)

  const ct = fromHex(vec.sample_ct_initiator)
  const pt = await rRecv.decryptWithAd(new Uint8Array(0), ct)
  assert.strictEqual(new TextDecoder().decode(pt), vec.sample_plaintext)
})

test('interop: JS can decrypt Go-generated ciphertext (responder->initiator)', async () => {
  // k2 is responder->initiator key; initiator uses k2 to decrypt
  const k2 = fromHex(vec.k2)
  const iRecv = new CipherState(k2)

  const ct = fromHex(vec.sample_ct_responder)
  const pt = await iRecv.decryptWithAd(new Uint8Array(0), ct)
  assert.strictEqual(new TextDecoder().decode(pt), vec.sample_plaintext)
})

test('interop: JS derives same k1/k2 from same fixed keys (PRIMARY)', async () => {
  // This is the primary interop gate. JS must reproduce the full deterministic
  // transcript from the Go-generated vector, not just decrypt ciphertext with
  // already-derived keys.
  const { importPrivateKeyFromScalar } = await import('./keys.js')
  const { NoiseXX } = await import('./index.js')

  const iEPriv = await importPrivateKeyFromScalar(fromHex(vec.initiator_ephemeral_priv))
  const rEPriv = await importPrivateKeyFromScalar(fromHex(vec.responder_ephemeral_priv))
  const iSPriv = await importPrivateKeyFromScalar(fromHex(vec.initiator_static_priv))
  const rSPriv = await importPrivateKeyFromScalar(fromHex(vec.responder_static_priv))

  const initiator = await NoiseXX.createInitiator({ staticKeypair: iSPriv, ephemeralKeypair: iEPriv })
  const responder = await NoiseXX.createResponder(rSPriv.privateKey, rSPriv.publicKeyBytes, { ephemeralKeypair: rEPriv })

  // Message 1: initiator -> responder (65 bytes)
  const msg1 = await initiator.writeMessage1()
  assert.strictEqual(Buffer.from(msg1).toString('hex'), vec.msg1, 'msg1 transcript mismatch')
  await responder.readMessage1(msg1)

  // Message 2: responder -> initiator (162 bytes)
  const msg2 = await responder.writeMessage2()
  assert.strictEqual(Buffer.from(msg2).toString('hex'), vec.msg2, 'msg2 transcript mismatch')
  await initiator.readMessage2(msg2)

  // Message 3: initiator -> responder (97 bytes)
  const msg3 = await initiator.writeMessage3()
  assert.strictEqual(Buffer.from(msg3).toString('hex'), vec.msg3, 'msg3 transcript mismatch')
  await responder.readMessage3(msg3)

  // Split and verify derived keys match Go exactly
  const [iSend, iRecv] = await initiator.split()
  const [rSend, rRecv] = await responder.split()

  // initiator sends on k1, receives on k2
  assert.strictEqual(Buffer.from(iSend.keyBytes).toString('hex'), vec.k1, 'k1 mismatch (initiator send)')
  assert.strictEqual(Buffer.from(iRecv.keyBytes).toString('hex'), vec.k2, 'k2 mismatch (initiator recv)')

  // responder sends on k2, receives on k1
  assert.strictEqual(Buffer.from(rRecv.keyBytes).toString('hex'), vec.k1, 'k1 mismatch (responder recv)')
  assert.strictEqual(Buffer.from(rSend.keyBytes).toString('hex'), vec.k2, 'k2 mismatch (responder send)')
})